### PR TITLE
[Based on transaction extension 3685] make transaction extension pipeline not rejecting not signed origins

### PR DESCRIPTION
**Pull request by gui1117**
Based on https://github.com/paritytech/polkadot-sdk/pull/3685

Update transaction extensions so that not signed origins are passing throught and correctly refunded.
* payments transaction extensions is done with refund.
  Ideally the transaction extension should be an option, so that we don't need to refund and instead we don't charge when the transaction extension is not used.
* `PrevalidateAttests` is refactor to have the weight at the correct place.
* `CheckNonce` is refunded for when nonce is not checked.
  Same ideally `Nonce` should be `Option<Nonce>`, so that we can have proper weight ahead of time and not need to do refund.
* bridges transaction extensions, allow not signed origins. But benchmark and weight will be done in another PR

I'm thinking we should be able to support multiple version of transaction extension at the same time in the runtime. So we can easily extend the capability of the transaction extensions.
Then we can have optional transaction payment and optional check nonce easily.

@georgepisaltu

### CODE REVIEW

**Review by georgepisaltu — COMMENTED** [review]
Looks good so far, let's add a test for each of the transaction payment extensions which checks the `NoCharge` behavior and asserts the full weight is refunded.

However, we should not merge this change together with phase 1 of extrinsic horizon (Transaction Extension), at least the way it is now.

This is because from the moment the transaction extension PR is merged, general transactions can be constructed. In the current verison of the Transaction Extension PR there is no `DenyNone` or custom `Applyable` logic to deny transactions with a `None` origin.

We need to either:
1. change the behavior in the Transaction Extension PR to disallow transactions with no origin
2. wait for the phase 2 PR to be completed to merge this (in other words, develop this PR further with the other changes)

**Comment by gui1117** [comment]
> We need to either:
> 
>     1. change the behavior in the Transaction Extension PR to disallow transactions with no origin
> 
>     2. wait for the phase 2 PR to be completed to merge this (in other words, develop this PR further with the other changes)

I would choose 1.

I was thinking with 2. that `ValidateUnsigned` will only be deprecated not removed.
And there is still inherent which are bare transaction which dispatch a `None` origin. Those transaction must be denied in the transaction extension. Either `DenyNone` or in the generic `DispatchTransaction` implementation in sp-runtime.

maybe something like this:
```diff
--- a/substrate/primitives/runtime/src/traits/transaction_extension/dispatch_transaction.rs
+++ b/substrate/primitives/runtime/src/traits/transaction_extension/dispatch_transaction.rs
@@ -89,7 +89,11 @@ impl<T: TransactionExtension<Call>, Call: Dispatchable + Encode> DispatchTransac
                info: &DispatchInfoOf<Call>,
                len: usize,
        ) -> Result<(ValidTransaction, T::Val, Self::Origin), TransactionValidityError> {
-               self.validate(origin, call, info, len, self.implicit()?, call)
+               let res = self.validate(origin, call, info, len, self.implicit()?, call)?;
+
+               ensure(!res.2.is_none(), ...);
+
+               res
        }
        fn validate_and_prepare(
                self,
```

and we can just remove `DenyNone`, people can implement it themselves if they have their own implementation of Extrinsic.

**Comment by gui1117** [comment]
bot bench-all pallet --pallet claims

**Comment by gui1117** [comment]
bot bench polkadot-pallet --pallet=claims

File: substrate/frame/transaction-payment/src/lib.rs [hunk 1]
```diff
@@ -850,28 +849,60 @@ impl<T: Config> core::fmt::Debug for ChargeTransactionPayment<T> {
...
+		// imbalance resulting from withdrawing the fee
+		imbalance: <<T as Config>::OnChargeTransaction as OnChargeTransaction<T>>::LiquidityInfo,
+	},
+	NoCharge {
+		// weight used by the extension
+		weight: Weight,
+	},
+}
+
+impl<T: Config> core::fmt::Debug for Pre<T> {
```
Comment (georgepisaltu) [inline substrate/frame/transaction-payment/src/lib.rs:881]: Why does `Pre` need to be `Debug`?

**Review by georgepisaltu — APPROVED** [review]

File: substrate/frame/transaction-payment/src/lib.rs (hunk 1 above)
Comment (gui1117) [reply substrate/frame/transaction-payment/src/lib.rs:881]: actually only for some tests.

**Pull request merged**